### PR TITLE
Add `Char#titlecase` for correct mixed-case transformations

### LINKED ISSUE #13533: `String#titleize` and `#capitalize` ignore Unicode `Titlecase_Mapping` property
Most characters that appear at the beginning of a string are converted into their uppercase forms when `String#capitalize` is called. A small number of characters, however, have a distinct titlecase form:

```crystal
# U+01F1 LATIN CAPITAL LETTER DZ
# U+01F2 LATIN CAPITAL LETTER D WITH SMALL LETTER Z
# U+01F3 LATIN SMALL LETTER DZ
# all of them should become U+01F2
"Ǳ".capitalize # => "Ǳ"
"ǲ".capitalize # => "ǲ"
"ǳ".capitalize # => "Ǳ"

# should become U+0046 U+0069
"ﬁ".capitalize # => "ﬁ"

# the entire Georgian Mkhedruli script (U+10D0-U+10FF) is considered lowercase,
# but should titlecase to itself rather than its uppercase (Mtavruli) block (U+1C90-U+1CBF)
"ა".capitalize # => "Ა"
```

Crystal should respect this. The same goes for `#titleize` (and probably `#camelcase` too).

### PR DESCRIPTION
**Pull request by HertzDevil**
Fixes #13533. The name `#titlecase` was specifically made to be different from `#titleize` because only the latter deals with word boundaries.

The full list of affected characters are:

<details>

| Character | | Titlecase | | Uppercase | |
|-|-|-|-|-|-|
| ß | U+00DF | Ss | U+0053 U+0073 | SS | U+0053 U+0053 |
| Ǆ | U+01C4 | ǅ | U+01C5 | Ǆ | U+01C4 |
| ǅ | U+01C5 | ǅ | U+01C5 | ǅ | U+01C5 |
| ǆ | U+01C6 | ǅ | U+01C5 | Ǆ | U+01C4 |
| Ǉ | U+01C7 | ǈ | U+01C8 | Ǉ | U+01C7 |
| ǈ | U+01C8 | ǈ | U+01C8 | ǈ | U+01C8 |
| ǉ | U+01C9 | ǈ | U+01C8 | Ǉ | U+01C7 |
| Ǌ | U+01CA | ǋ | U+01CB | Ǌ | U+01CA |
| ǋ | U+01CB | ǋ | U+01CB | ǋ | U+01CB |
| ǌ | U+01CC | ǋ | U+01CB | Ǌ | U+01CA |
| Ǳ | U+01F1 | ǲ | U+01F2 | Ǳ | U+01F1 |
| ǲ | U+01F2 | ǲ | U+01F2 | ǲ | U+01F2 |
| ǳ | U+01F3 | ǲ | U+01F2 | Ǳ | U+01F1 |
| և | U+0587 | Եւ | U+0535 U+0582 | ԵՒ | U+0535 U+0552 |
| ა | U+10D0 | ა | U+10D0 | Ა | U+1C90 |
| ბ | U+10D1 | ბ | U+10D1 | Ბ | U+1C91 |
| გ | U+10D2 | გ | U+10D2 | Გ | U+1C92 |
| დ | U+10D3 | დ | U+10D3 | Დ | U+1C93 |
| ე | U+10D4 | ე | U+10D4 | Ე | U+1C94 |
| ვ | U+10D5 | ვ | U+10D5 | Ვ | U+1C95 |
| ზ | U+10D6 | ზ | U+10D6 | Ზ | U+1C96 |
| თ | U+10D7 | თ | U+10D7 | Თ | U+1C97 |
| ი | U+10D8 | ი | U+10D8 | Ი | U+1C98 |
| კ | U+10D9 | კ | U+10D9 | Კ | U+1C99 |
| ლ | U+10DA | ლ | U+10DA | Ლ | U+1C9A |
| მ | U+10DB | მ | U+10DB | Მ | U+1C9B |
| ნ | U+10DC | ნ | U+10DC | Ნ | U+1C9C |
| ო | U+10DD | ო | U+10DD | Ო | U+1C9D |
| პ | U+10DE | პ | U+10DE | Პ | U+1C9E |
| ჟ | U+10DF | ჟ | U+10DF | Ჟ | U+1C9F |
| რ | U+10E0 | რ | U+10E0 | Რ | U+1CA0 |
| ს | U+10E1 | ს | U+10E1 | Ს | U+1CA1 |
| ტ | U+10E2 | ტ | U+10E2 | Ტ | U+1CA2 |
| უ | U+10E3 | უ | U+10E3 | Უ | U+1CA3 |
| ფ | U+10E4 | ფ | U+10E4 | Ფ | U+1CA4 |
| ქ | U+10E5 | ქ | U+10E5 | Ქ | U+1CA5 |
| ღ | U+10E6 | ღ | U+10E6 | Ღ | U+1CA6 |
| ყ | U+10E7 | ყ | U+10E7 | Ყ | U+1CA7 |
| შ | U+10E8 | შ | U+10E8 | Შ | U+1CA8 |
| ჩ | U+10E9 | ჩ | U+10E9 | Ჩ | U+1CA9 |
| ც | U+10EA | ც | U+10EA | Ც | U+1CAA |
| ძ | U+10EB | ძ | U+10EB | Ძ | U+1CAB |
| წ | U+10EC | წ | U+10EC | Წ | U+1CAC |
| ჭ | U+10ED | ჭ | U+10ED | Ჭ | U+1CAD |
| ხ | U+10EE | ხ | U+10EE | Ხ | U+1CAE |
| ჯ | U+10EF | ჯ | U+10EF | Ჯ | U+1CAF |
| ჰ | U+10F0 | ჰ | U+10F0 | Ჰ | U+1CB0 |
| ჱ | U+10F1 | ჱ | U+10F1 | Ჱ | U+1CB1 |
| ჲ | U+10F2 | ჲ | U+10F2 | Ჲ | U+1CB2 |
| ჳ | U+10F3 | ჳ | U+10F3 | Ჳ | U+1CB3 |
| ჴ | U+10F4 | ჴ | U+10F4 | Ჴ | U+1CB4 |
| ჵ | U+10F5 | ჵ | U+10F5 | Ჵ | U+1CB5 |
| ჶ | U+10F6 | ჶ | U+10F6 | Ჶ | U+1CB6 |
| ჷ | U+10F7 | ჷ | U+10F7 | Ჷ | U+1CB7 |
| ჸ | U+10F8 | ჸ | U+10F8 | Ჸ | U+1CB8 |
| ჹ | U+10F9 | ჹ | U+10F9 | Ჹ | U+1CB9 |
| ჺ | U+10FA | ჺ | U+10FA | Ჺ | U+1CBA |
| ჽ | U+10FD | ჽ | U+10FD | Ჽ | U+1CBD |
| ჾ | U+10FE | ჾ | U+10FE | Ჾ | U+1CBE |
| ჿ | U+10FF | ჿ | U+10FF | Ჿ | U+1CBF |
| ᾲ | U+1FB2 | Ὰͅ | U+1FBA U+0345 | ᾺΙ | U+1FBA U+0399 |
| ᾴ | U+1FB4 | Άͅ | U+0386 U+0345 | ΆΙ | U+0386 U+0399 |
| ᾷ | U+1FB7 | ᾼ͂ | U+0391 U+0342 U+0345 | Α͂Ι | U+0391 U+0342 U+0399 |
| ῂ | U+1FC2 | Ὴͅ | U+1FCA U+0345 | ῊΙ | U+1FCA U+0399 |
| ῄ | U+1FC4 | Ήͅ | U+0389 U+0345 | ΉΙ | U+0389 U+0399 |
| ῇ | U+1FC7 | ῌ͂ | U+0397 U+0342 U+0345 | Η͂Ι | U+0397 U+0342 U+0399 |
| ῲ | U+1FF2 | Ὼͅ | U+1FFA U+0345 | ῺΙ | U+1FFA U+0399 |
| ῴ | U+1FF4 | Ώͅ | U+038F U+0345 | ΏΙ | U+038F U+0399 |
| ῷ | U+1FF7 | ῼ͂ | U+03A9 U+0342 U+0345 | Ω͂Ι | U+03A9 U+0342 U+0399 |
| ﬀ | U+FB00 | Ff | U+0046 U+0066 | FF | U+0046 U+0046 |
| ﬁ | U+FB01 | Fi | U+0046 U+0069 | FI | U+0046 U+0049 |
| ﬂ | U+FB02 | Fl | U+0046 U+006C | FL | U+0046 U+004C |
| ﬃ | U+FB03 | Ffi | U+0046 U+0066 U+0069 | FFI | U+0046 U+0046 U+0049 |
| ﬄ | U+FB04 | Ffl | U+0046 U+0066 U+006C | FFL | U+0046 U+0046 U+004C |
| ﬅ | U+FB05 | St | U+0053 U+0074 | ST | U+0053 U+0054 |
| ﬆ | U+FB06 | St | U+0053 U+0074 | ST | U+0053 U+0054 |
| ﬓ | U+FB13 | Մն | U+0544 U+0576 | ՄՆ | U+0544 U+0546 |
| ﬔ | U+FB14 | Մե | U+0544 U+0565 | ՄԵ | U+0544 U+0535 |
| ﬕ | U+FB15 | Մի | U+0544 U+056B | ՄԻ | U+0544 U+053B |
| ﬖ | U+FB16 | Վն | U+054E U+0576 | ՎՆ | U+054E U+0546 |
| ﬗ | U+FB17 | Մխ | U+0544 U+056D | ՄԽ | U+0544 U+053D |

</details>

### CODE REVIEW

File: scripts/generate_unicode_data.cr [hunk 1]
```diff
@@ -282,6 +295,11 @@ upcase_ranges.select! { |r| r.delta != -1 }
 
 alternate_ranges = alternate_ranges(downcase_one_ranges)
 
+special_cases_downcase.sort_by! &.codepoint
+special_cases_upcase.sort_by! &.codepoint
+special_cases_titlecase.reject! { |v| special_cases_upcase.includes?(v) }
```
Comment (Sija) [inline scripts/generate_unicode_data.cr:300]:
```suggestion
special_cases_titlecase.reject! &.in?(special_cases_upcase)
```

File: spec/std/string_spec.cr [hunk 2]
```diff
@@ -693,47 +693,45 @@ describe "String" do
...
       "\xB5!\xE0\xC1\xB5?".capitalize.should eq("\xB5!\xE0\xC1\xB5?")
     end
 
-    describe "with IO" do
-      it { String.build { |io| "HELLO!".capitalize io }.should eq "Hello!" }
-      it { String.build { |io| "HELLO MAN!".capitalize io }.should eq "Hello man!" }
-      it { String.build { |io| "".capitalize io }.should be_empty }
-      it { String.build { |io| "ﬄİ".capitalize io }.should eq "FFLi̇" }
-      it { String.build { |io| "iO".capitalize io, Unicode::CaseOptions::Turkic }.should eq "İo" }
+    pending "IO overload with invalid code units" do
```
Comment (beta-ziliani) [inline spec/std/string_spec.cr:709]: Why is this pending? What should it do?

Comment (HertzDevil) [reply spec/std/string_spec.cr:709]: The `IO` overloads should convert the invalid UTF-8 byte sequences into U+FFFD despite #12108, as they must not inadvertently form any valid sequences in whatever the target `IO`'s encoding is. Now I think that can be expressed in `assert_prints` itself

File: src/char.cr [hunk 3]
```diff
@@ -474,6 +488,49 @@ struct Char
...
+    Unicode.titlecase(self, options)
+  end
+
+  # Yields each char for the titlecase equivalent of this char.
+  #
+  # Usually this is equivalent to `#upcase`, but a few precomposed characters
+  # consisting of multiple letters may yield a different character sequence
+  # where only the first letter is uppercase and the rest lowercase.
+  #
+  # This method takes into account the possibility that an upcase
```
Comment (straight-shoota) [inline src/char.cr:521]:
```suggestion
  # This method takes into account the possibility that an titlecase
```

**Review by straight-shoota — APPROVED** [review]

**Pull request merged**